### PR TITLE
Add ability to read/write LazyVector via VectorSaver utility

### PR DESCRIPTION
**Pull request by bikramSingh91**
This patch adds the ability to serialize a lazy vector via the
VectorSaver utility. The lazy vector saved has the header
(size, type) and the loaded vector (if it was loaded at the
time of serialization). The idea here is to recreate the state
of the loaded vector when expression evaluation errored out
(which initiated its serialization). This would help in quickly
reproducing the error as the lazy vector would end up in the
same state if the same steps are executed that resulted in an
error in the first place. However this also precludes it from
being used in any other context like trying to test a fix.

TestPlan:
Added unit test

### CODE REVIEW

**Review by mbasmanova — COMMENTED** [review]
@bikramSingh91 Would you update PR description to explain how Lazy vectors are persisted and restored? The PR also has some changes to VectorFuzzer. It would be nice to describe these changes as well. Finally, would be great to include updates to the documentation: https://facebookincubator.github.io/velox/develop/debugging/vector-saver.html

File: velox/vector/VectorSaver.cpp [hunk 1]
```diff
@@ -508,6 +513,56 @@ VectorPtr readMapVector(
   return std::make_shared<MapVector>(
       pool, type, nulls, size, offsets, sizes, keys, values);
 }
+
+void writeLazyVector(const BaseVector& vector, std::ostream& out) {
+  auto lazyVector = dynamic_cast<const LazyVector*>(&vector);
+  // Nulls buffer.
```
Comment (mbasmanova) [inline velox/vector/VectorSaver.cpp:519]: Comment needs updating.

File: velox/vector/VectorSaver.cpp [hunk 2]
```diff
@@ -508,6 +513,56 @@ VectorPtr readMapVector(
...
+  bool isLoaded = lazyVector->isLoaded();
+  write<bool>(isLoaded, out);
+
+  // loaded vector.
+  if (isLoaded) {
```
Comment (mbasmanova) [inline velox/vector/VectorSaver.cpp:524]: What happens if the vector is not loaded? Looks like we are not persisting anything in this case. How will it be restored?

File: velox/vector/fuzzer/VectorFuzzer.cpp [hunk 3]
```diff
@@ -659,6 +659,19 @@ RowVectorPtr VectorFuzzer::fuzzRowChildrenToLazy(RowVectorPtr rowVector) {
       std::move(children));
 }
 
+SelectivityVector VectorFuzzer::fuzzSelectivity(
```
Comment (mbasmanova) [inline velox/vector/fuzzer/VectorFuzzer.cpp:662]: This change seems unrelated. Should it be in a different PR?

File: velox/vector/VectorSaver.cpp (hunk 2 above)
Comment (bikramSingh91) [reply velox/vector/VectorSaver.cpp:524]: The restored lazy vector will only have the format and size and would fail if it an attempt is made to load it. the idea here is to be able to repro an issue, so if the repro happened before the lazy vector got materialized, this will let us create a placeholder for it with the right format instead of skipping vector serialization altogether and allow for a quick repro.

File: velox/vector/fuzzer/VectorFuzzer.cpp (hunk 3 above)
Comment (bikramSingh91) [reply velox/vector/fuzzer/VectorFuzzer.cpp:662]: will move this out in separate PR

File: velox/vector/VectorSaver.cpp (hunk 2 above)
Comment (mbasmanova) [reply velox/vector/VectorSaver.cpp:524]: @bikramSingh91 Thank you for explaining. Makes sense. It would be great to document this as it is not intuitive.

File: velox/docs/develop/debugging/vector-saver.rst [hunk 4]
```diff
@@ -176,3 +176,21 @@ Dictionary Vector
 * Nulls buffer (if present).
 * Indices buffer.
 * Base vector.
+
+Lazy Vector
```
Comment (mbasmanova) [inline velox/docs/develop/debugging/vector-saver.rst:180]: > * 0 - FLAT, 1 - CONSTANT, 2 - DICTIONARY

Encoding section above needs to be updated to add LAZY encoding.

**Review by mbasmanova — APPROVED** [review]
@bikramSingh91 Very creative solution. Thank you for enhancing VectorSaver to support LazyVector.
